AutomatticTracks/src/androidTest/java/com/automattic/android/tracks/CrashLogging/SendEventsToCrashLoggerService.kt

```diff
@@ -0,0 +1,61 @@
+package com.automattic.android.tracks.crashlogging
+
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.platform.app.InstrumentationRegistry
+import com.automattic.android.tracks.TracksUser
+import com.automattic.android.tracks.fakes.FakeDataProvider
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class SendEventsToCrashLoggerService {
```

Comment:
The idea for this "service" is to have some mechanism for frequent end-to-end check of the Sentry integration. It's not a test, even if it's in `androidTest` directory.

This way we're to run a lot of tests on test Sentry project. I believe this will be helpful also for future work.

## Demo


https://user-images.githubusercontent.com/5845095/111218209-b91b4600-85d6-11eb-92e4-8760d9a23273.mp4

Reply:
I am not sure how I feel about this. I can understand the idea, but it feels odd to me to do it like this. Since I can't provide a meaningful counterpoint not to do it, I guess I'll just leave a couple suggestions in case we keep it:

* The `Service` postfix is [a reserved keyword in Android](https://developer.android.com/guide/components/services) and it really doesn't describe what this class does. So, I suggest using a more descriptive without the `Service` in it. 
* We should document how to use this class and explain why it's in the project in the first place.
* `waitForEventToBeSent` doesn't seem to do anything. I understand it's communicating that the event will take time to appear, but I don't think it should actually cause a delayed finish unless we are taking an action after the delay. We should just have the delay communicated in the documentation I mentioned.
* We shouldn't run these tests in CI. We have a quota in Sentry and they really don't provide any value unless we are actively testing the Sentry integration.

---

While typing the suggestions, I realized why I feel uneasy about this. I think this makes a lot more sense as an example app instead of an instrumentation test. We don't verify anything in the tests, so they provide no value as an automated test. We could potentially add a way to check if the Sentry event is there (using their API), but I don't see enough value in that.

Instead, I think we can have a very simple app within this repo that simply adds Tracks library to it and have buttons in the UI that calls these different methods. I can definitely get behind something like that.

Reply:
> The Service postfix is a reserved keyword in Android

Sure, I'll rename that to not cause any confusion

> We should document how to use this class and explain why it's in the project in the first place.

What kind of documentation do you think of? More live Javadoc like, or some instructions on P2 or README on Github? What would suit here the best?

> `waitForEventToBeSent` doesn't seem to do anything (...) We should just have the delay communicated in the documentation I mentioned.

`waitForEventToBeSent` does something and it's locking a main thread to not allow the test app to close immediately after test execution. Otherwise, the events would not be sent because test app will close before HTTP request from Sentry will have a chance to be sent. Mentioning this delay in documentation won't change anything.

> We shouldn't run these tests in CI

And we don't do this, do we? Current CI task is `./gradlew --stacktrace testRelease` and it does not have any dependencies on `connectedAndroidTest` which runs instrumented tests on Android. If we would run this test on CI, even in this build it would fail as there's no setup for the emulator to run on CI.

---

>  I think this makes a lot more sense as an example app instead of an instrumentation test.

I get your point that this class is not a test and I don't want to make this a test. 

I'd opt against having sample app for testing Sentry. Sample app causes a lot of additional overhead like keeping dependencies up to date or writing UI and platform code. Also, typical testing flow will require more effort with rebuilding the app and manually pressing UI buttons. I think this mechanism I've proposed is faster to work with and easier to maintain. To test if everything works fine after changes, developers will only need to run methods from one class.

Do those additional arguments convince you to leave this mechanism (I intentionally avoid word `test`)?

Reply:
> Could you please elaborate on this part "Mentioning this delay in documentation won't change anything."

What I meant by writing this is that this method is not about "communicating that the event will take time to appear". It's a vital part of sending a message as tests run on the test app (I mean: created for tests) and for every test execution we've got a new run of the app. After each test is finished app is closed so without locking the main thread (=pausing test for a few seconds) Sentry won't have time to be sent.

Rephrasing this statement to be more descriptive: "Knowing about delay won't resolve any issue here. It's not about informing developer about something - this pause is a must". So this is what I meant: just knowing that there's a delay won't happen that this will work.

> every class which is annotated with "RunWith" is considered as a test by Gradle

You're right, by saying "this class is not a test" I meant that it's not a test in a formal way, not that this is not a test by Gradle standards. 

> We might want to consider adding "@ignore" annotation and a comment explaining that this test is disabled

That's a good idea, I'm starting to drifting into a sample app solution though (more on that in the latter part of this comment) so it might not be needed.

---

Thank you for your input on that ""service"" idea. I can see that this idea raises a lot of concerns so eventually, it might not be a good fit.

Although I'm not in love with the idea of the sample app (I just have aversion to buttons :) ) I'll prepare one to avoid any confusion now and in the future.

Reply:
If it's okay for you, I've created an issue #69 and I will work on creating a sample app on a separate pull request to not shadow changes regarding Sentry bump.

I think it's okay to leave two options: automated sending of events using Android's instrumentation tests framework and a sample app.

> We might want to consider adding "@ignore" annotation and a comment explaining that this test is disabled

After reconsidering I'd like to suggest not adding `@Ignore` to this "test". A scenario when it'll run alongside "real instrumentation tests" is not a case for now (and I don't suspect it'll change soon) and if it will happen - adding `@Ignore` is a little effort. Is that okay or you'd prefer to add `@Ignore` now?

I use this method quite often and it'd be a bummer to have to add and remove `@Ignore` on every interaction with git.

Reply:
> Rephrasing this statement to be more descriptive: "Knowing about delay won't resolve any issue here. It's not about informing developer about something - this pause is a must". So this is what I meant: just knowing that there's a delay won't happen that this will work.

I should have read the comments more carefully 😞 . Thanks for the explanation!

> I think it's okay to leave two options: automated sending of events using Android's instrumentation tests framework and a sample app.
> After reconsidering I'd like to suggest not adding @Ignore to this "test". A scenario when it'll run alongside "real instrumentation tests" is not a case for now (and I don't suspect it'll change soon) and if it will happen - adding @Ignore is a little effort. Is that okay or you'd prefer to add @Ignore now?
I use this method quite often and it'd be a bummer to have to add and remove @Ignore on every interaction with git.

I'm just a bit worried that whoever starts running tests on a CI probably won't even know that they should prevent these specific tests from running. I'm also worried about the precedens - if we use it here, it'll be much easier to use the same approach again in other repos. 
The fact that it's a bit cumbersome is imo the price for using a tool which is not meant to be used for "tests" which require a human interaction. However, I'm wondering if we could perhaps have some setting in gradle.properties instead of using @Ignore. Wdyt @wzieba @oguzkocer ?

Reply:
> The fact that it's a bit cumbersome is imo the price for using a tool which is not meant to be used for "tests" which require a human interaction

I agree with this and previous points - still, I believe it's a very helpful tool to test this library e2e.

I think I've found a compromise (IMO quite neat!) in 8d74cf24682ee0cb9c9c33d79bc9af8f126d3cde . Current behavior:
- Running `./gradlew connectedAndroidTest` results with `No tests found`
- Running `SendEventsToSentry` from Android Studio will run this test as we explicitly ask to run this class

WDYT? Does it work for you this way?

Reply:
TIL about `'notClass'` argument! LGTM ;)

Reply:
> TIL about 'notClass' argument!

Me too! From this article if you're interested diving into more details: https://medium.com/stepstone-tech/exploring-androidjunitrunner-filtering-options-df26d30b4f60